### PR TITLE
TEST: Add unit tests for `MemcachedConnection.handleCacheNodesChange()` method in replication mode.

### PR DESCRIPTION
**Pull request by uhm0311**
### 🔗 Related Issue

- https://github.com/jam2in/arcus-works/issues/598
- https://github.com/jam2in/arcus-works/issues/620
- https://github.com/jam2in/arcus-works/issues/637

### ⌨️ What I did

- 복제 관련 코드 리팩토링 진행 시 기존 로직과 달라진 경우를 로컬 개발 단계에서 바로 확인할 수 있도록 단위 테스트를 추가합니다.

### CODE REVIEW

**Review by brido4125 — APPROVED** [review]

File: src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    SocketAddress masterAddr = master.getSocketAddress();
+    SocketAddress slaveAddr = slave.getSocketAddress();
+    assertEquals("{g0 M 127.0.0.2:11211}", masterAddr.toString());
+    assertEquals("{g0 S 127.0.0.1:11211}", slaveAddr.toString());
+  }
+
+  @Test
+  void testHandleCacheNodesChange_failover() throws IOException {
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses(
+            "g0^M^127.0.0.1:11211,g0^S^127.0.0.2:11211"));
```
Comment (oliviarla) [inline src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java:114]: master노드에 장애가 발생해 failover로 제외되는 케이스도 추가하면 어떤가요?

Comment (uhm0311) [reply src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java:140]: master에 장애가 발생하는 경우, 둘 다 발생하는 경우도 추가했습니다.

**Review by oliviarla — APPROVED** [review]

File: src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+  void testHandleCacheNodesChange_switchover() throws IOException {
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses(
+            "g0^M^127.0.0.1:11211,g0^S^127.0.0.2:11211"));
+    conn.handleCacheNodesChange();
+
+    assertEquals(2, locator.getAll().size());
+    assertEquals(1, locator.getAllGroups().size());
+
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses(
+            "g0^M^127.0.0.2:11211,g0^S^127.0.0.1:11211"));
```
Comment (jhpark816) [inline src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java:90]: swithover 경우 ZK 이벤트는 아래 순서로 변경하게 되므로, 2회의 변경 처리이면 좋겠습니다.
- switchover 수행 전
  - "g0^M^127.0.0.1:11211,g0^S^127.0.0.2:11211"
- switchover 수행 후 (기존 master znode 변경)
  - "g0^S^127.0.0.1:11211,g0^S^127.0.0.2:11211"
- switchover 수행 후 (기존 slave znode 변경)
  - "g0^S^127.0.0.1:11211,g0^M^127.0.0.2:11212"


위에서 invalid group 테스트를 하기 때문에
 invalid group 테스트는 따로 하지 않아도 됩니다.

File: src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+  @Test
+  void testHandleCacheNodesChange_failover_master() throws IOException {
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses(
+            "g0^M^127.0.0.1:11211,g0^S^127.0.0.2:11211"));
+    conn.handleCacheNodesChange();
+
+    assertEquals(2, locator.getAll().size());
+    assertEquals(1, locator.getAllGroups().size());
+
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses("g0^M^127.0.0.2:11211"));
```
Comment (jhpark816) [inline src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java:120]: 여기도 2회의 변경이 발생합니다.

- T0) "g0^M^127.0.0.1:11211,g0^S^127.0.0.2:11211"
- T1) "g0^S^127.0.0.2:11211"
- T2) "g0^M^127.0.0.2:11211"

File: src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+  @Test
+  void testHandleCacheNodesChange_exchange_group_name_with_switchover() throws IOException {
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses(
+            "g0^M^127.0.0.1:11211,g0^S^127.0.0.2:11211,g1^M^127.0.0.3:11211,g1^S^127.0.0.4:11211"));
+    conn.handleCacheNodesChange();
+    assertEquals(4, locator.getAll().size());
+    assertEquals(2, locator.getAllGroups().size());
+
+    conn.setCacheNodesChange(ArcusReplNodeAddress.getAddresses(
+            "g0^M^127.0.0.4:11211,g0^S^127.0.0.3:11211,g1^M^127.0.0.2:11211,g1^S^127.0.0.1:11211"));
```
Comment (jhpark816) [inline src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java:305]: 이렇게 switchover 발생하는 경우가 없으니, 이러한 테스트는 없어도 되지 않나 생각합니다.

**Review by jhpark816 — COMMENTED** [review]
리뷰 완료

File: src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java (hunk 4 above)
Comment (uhm0311) [reply src/test/manual/net/spy/memcached/MemcachedConnectionReplTest.java]: 일부러 만든 케이스이므로 메소드 이름에 edge case를 넣는 것으로 합의하였습니다.

**Review by jhpark816 — APPROVED** [review]
